js/autotrack/common.ts

```diff
@@ -1,19 +1,59 @@
 import * as _ from 'lodash';
 
+import { Component as ReactComponent } from 'React';
+import { Fiber as FiberNode } from 'react-reconciler';
 import { extractProps } from '../util/extractProps';
 import { BASE_HEAP_IGNORE_PROPS, getNextHeapIgnoreProps } from './heapIgnore';
 import { builtinPropExtractorConfig } from '../propExtractorConfig';
 
+// Base properties for autotracked events.
+interface AutotrackProps {
+  touchableHierarchy: string;
+  targetText?: string;
+}
+
+// The type definition of 'Component' from '@types/react' doesn't include the internal
+// '_reactInternalFiber' property, so create our own 'Component' type that includes this prop.
+interface Component extends ReactComponent {
+  _reactInternalFiber: FiberNode;
+}
+
+interface HeapIgnoreProps {
+  ignoreInteraction?: boolean;
+  ignoreInnerHierarchy?: boolean;
+  ignoreAllProps?: boolean;
+  // :TODO: (jmtaber129): Add 'ignoreSpecificProps'.
+  ignoreTargetText?: boolean;
+}
+
+// The result of traversing up and back down a component's hierarchy.
+interface HierarchyResult {
+  hierarchy: string;
+  heapIgnoreProps: HeapIgnoreProps;
+}
+
+// The representation of a component in a Fiber hierarchy.
+interface ComponentHierarchyTraversalElement {
+  elementName: string;
+  fiberNode: FiberNode;
+  propsString: string;
+}
+
 // Returns an object containing a base set of component properties if we're not ignoring the
 // full interaction due to HeapIgnore.
 // Returns null if we're ignoring the full interaction due to HeapIgnore.
-export const getBaseComponentProps = componentThis => {
+export const getBaseComponentProps: (
+  component: Component
+) => AutotrackProps | null = componentThis => {
   // Get the hierarchy traversal from root to target component, then get the actual hierarchy from
   // the traversal representation.
-  const touchableHierarchyTraversal = getComponentHierarchyTraversal(
+  const touchableHierarchyTraversal: ComponentHierarchyTraversalElement[] = getComponentHierarchyTraversal(
```

Comment:
Do you have to do this explicitly here?  Shouldn't it get picked up by the return type of the func?

Reply:
This _could_ be implicit, but I personally prefer explicitly specifying the type here (since the type isn't immediately apparent from surrounding code).

I guess we don't have a style guide that talks about this, but to me, leaving off the type here would be akin to using the `auto` type from other languages, which I've been taught to avoid except with long and/or unnecessary type names.

Reply:
In the Heap project, we don't require explicit type annotation on assignments like this, but we do require explicit annotations on class property declarations and function return types. This is all enforced with `tslint`.

I think it's probably having a larger discussion around to develop some consensus. Personally, I prefer to use type inference in as many places as possible, but this is also means that code needs to be designed to be readable without accompanying type annotations. There is no hard and fast rule here, yet.

Reply:
Okay, with Luke's explanation of where explicit type annotations are required, along with being reminded of [TS strict mode](https://mariusschulz.com/blog/typescript-2-3-the-strict-compiler-option), I'm realizing that an explicit here is _not_ how I've preferred to handle type annotations in the past.

Switched this to be an implicit typing